alt-tab-macos/ui/PreferencesPanel.swift

```diff
@@ -1,110 +1,211 @@
-    var maxScreenUsage: NSTextView?
-    var windowPadding: NSTextView?
-    var cellPadding: NSTextView?
-    var cellBorderWidth: NSTextView?
-    var maxThumbnailsPerRow: NSTextView?
-    var iconSize: NSTextView?
-    var fontHeight: NSTextView?
-    var interItemPadding: NSTextView?
-    var tabKeyCode: NSTextView?
-    var windowDisplayDelay: NSTextView?
+class PreferencesPanel: NSPanel {
+    // ui: base layout
+    let panelWidth = CGFloat(400)
+    let panelHeight = CGFloat(400) // gets auto adjusted to content height
+    let panelPadding = CGFloat(40)
+    let panelWidthToLabelRatio = CGFloat(0.5)
+
+    // ui: preferences elements
+    var maxScreenUsage: NSSlider?
+    var maxThumbnailsPerRow: NSTextField?
+    var iconSize: NSSlider?
+    var fontHeight: NSSlider?
+    var tabKeyCode: NSTextField?
+    var windowDisplayDelay: NSTextField?
     var metaKey: NSPopUpButton?
     var theme: NSPopUpButton?
     var showOnScreen: NSPopUpButton?
-    var inputsMap = [NSTextView: String]()
+
+    var invisibleTextField: NSTextField? // default firstResponder and used for triggering of focus loose
 
     override init(contentRect: NSRect, styleMask style: StyleMask, backing backingStoreType: BackingStoreType, defer flag: Bool) {
-        super.init(contentRect: contentRect, styleMask: style, backing: backingStoreType, defer: flag)
+        let initialRect = NSRect(x: 0, y: 0, width: panelWidth, height: panelHeight)
+        super.init(contentRect: initialRect, styleMask: style, backing: backingStoreType, defer: flag)
         title = Application.name + " Preferences"
+        titlebarAppearsTransparent = true
         hidesOnDeactivate = false
-        contentView = makeGridView(makeLabelsAndInputs(), makeWarningLabel())
+        contentView = makeContentView()
+
+        // setup hidden element
+        invisibleTextField = NSTextField()
+        invisibleTextField?.isHidden = true
+        contentView?.subviews.append(invisibleTextField!)
+        self.initialFirstResponder = invisibleTextField
     }
 
-    private func makeLabelsAndInputs() -> [[NSView]] {
+    private func makeContentView() -> NSView {
+        let wrappingView = NSStackView(views: makePreferencesViews())
+        let contentView = NSView()
+        contentView.addSubview(wrappingView)
+
+        // visual setup
+        wrappingView.orientation = .vertical
+        wrappingView.alignment = .left
+        wrappingView.spacing = panelPadding * 0.3
+        wrappingView.topAnchor.constraint(equalTo: contentView.topAnchor, constant: panelPadding * 0.5).isActive = true
+        wrappingView.bottomAnchor.constraint(equalTo: contentView.bottomAnchor, constant: panelPadding * -0.5).isActive = true
+        wrappingView.leftAnchor.constraint(equalTo: contentView.leftAnchor, constant: panelPadding * 0.5).isActive = true
+        wrappingView.rightAnchor.constraint(equalTo: contentView.rightAnchor, constant: panelPadding * -0.5).isActive = true
```

Comment:
I started having way too many of these lines in the `Application` class. Then I extracted them in sub-methods to make it more readable at a high level. Then I realize that the people who made this UI framework are really thinking OOP all the way, and you are expected to have subclasses to customize your visual elements.

Also you can make the parallel with Angular/React and think that this `wrappingView`/`NSStackView` may be reuse in some other place one day, so storing it in its own file in the `ui` folder kind of make sense.

Once I realized the framework designers intent, I extracted a bunch of files/classes in the `ui` folder. I didn't do it for every single 2 liner UI element though. I think there is a balance between having too many files in that folder, and having too many blocks like this one here in bigger classes.

I think that one should get its own class

Note that I think I will do it when I have some time for `makeBackgroundView` and `makeCollectionView` in `ThumbnailsPanel`. They are similar candidates to should get their own subclasses I think

Reply:
Alright. Also a sub-directory structure could begin to make sense, like feature -> layer. ThumbnailPanel could get its own sub-directory and all the related classes.